Mathlib/LinearAlgebra/DirectSum/Finsupp.lean

```diff
@@ -24,36 +58,134 @@ open DirectSum
+      (Finsupp.mapRange.linearEquiv (TensorProduct.comm R _ _))
+
+/-- The tensor product of `M` and `i →₀ N` is linearly equivalent to `i →₀ M ⊗[R] N` -/
+noncomputable def Finsupp.lTensor :
+    M ⊗[R] (ι →₀ N) ≃ₗ[R] ι →₀ (M ⊗[R] N) :=
+  (TensorProduct.congr (LinearEquiv.refl R M) (finsuppLEquivDirectSum R N ι)).trans
+    ((TensorProduct.directSumRight R M (fun _ : ι => N)).trans
+      (finsuppLEquivDirectSum R (M ⊗[R] N) ι).symm)
+
+lemma Finsupp.lTensor_apply_tmul (m : M) (p : ι →₀ N) :
+    Finsupp.lTensor (m ⊗ₜ[R] p) = p.sum (fun i n ↦ Finsupp.single i (m ⊗ₜ[R] n)) := by
+  simp [Finsupp.lTensor]
+  conv_lhs => rw [← Finsupp.sum_single p]
+  rw [LinearEquiv.symm_apply_eq]
+  simp only [map_finsupp_sum]
+  simp only [finsuppLEquivDirectSum_single]
+  rw [← LinearEquiv.eq_symm_apply]
+  simp only [map_finsupp_sum]
+  simp only [directSumRight_symm_lof_tmul]
+  simp only [Finsupp.sum, tmul_sum]
+
+lemma Finsupp.lTensor_symm_apply_single (i : ι) (m : M) (n : N) :
+    Finsupp.rTensor.symm (Finsupp.single i (m ⊗ₜ[R] n)) =
+      Finsupp.single i m ⊗ₜ[R] n := by
+  simp [Finsupp.rTensor, Finsupp.lsum]
+
+variable {S : Type*} [CommSemiring S] [Algebra R S]
+  [Module S M] [IsScalarTower R S M]
+
+lemma Finsupp.rTensor_smul' (s : S) (t : (ι →₀ M) ⊗[R] N) :
+    Finsupp.rTensor (s • t) = s • Finsupp.rTensor t := by
+  induction t using TensorProduct.induction_on with
+  | zero => simp
+  | add x y hx hy =>
+    simp only [AddHom.toFun_eq_coe, coe_toAddHom, LinearEquiv.coe_coe, RingHom.id_apply] at hx hy ⊢
+    simp only [smul_add, map_add, hx, hy]
+  | tmul p n =>
+    simp only [TensorProduct.smul_tmul', rTensor_apply_tmul]
+    rw [Finsupp.smul_sum]
+    simp only [smul_single]
+    apply Finsupp.sum_smul_index'
+    simp
+
+/-- When `M` is also an `S`-module, then `Finsupp.rTensor R M N``
+  is an `S`-linear equiv -/
+noncomputable def Finsupp.rTensor' :
+    (ι →₀ M) ⊗[R] N ≃ₗ[S] ι →₀ (M ⊗[R] N) := {
+  Finsupp.rTensor with
+  map_smul' := Finsupp.rTensor_smul' }
+
+/- -- TODO : reprove using the existing heterobasic lemmas
+noncomputable example :
+    (ι →₀ M) ⊗[R] N ≃ₗ[S] ι →₀ (M ⊗[R] N) := by
+  have f : (⨁ (i₁ : ι), M) ⊗[R] N ≃ₗ[S] ⨁ (i : ι), M ⊗[R] N := sorry
+  exact (TensorProduct.AlgebraTensorModule.congr
+    (finsuppLEquivDirectSum S M ι) (LinearEquiv.refl R N)).trans
+    (f.trans (finsuppLEquivDirectSum S (M ⊗[R] N) ι).symm) -/
 
 open scoped Classical in
 -- `noncomputable` is a performance workaround for mathlib4#7103
 /-- The tensor product of `ι →₀ M` and `κ →₀ N` is linearly equivalent to `(ι × κ) →₀ (M ⊗ N)`. -/
-noncomputable def finsuppTensorFinsupp (R M N ι κ : Sort _) [CommRing R] [AddCommGroup M]
-    [Module R M] [AddCommGroup N] [Module R N] : (ι →₀ M) ⊗[R] (κ →₀ N) ≃ₗ[R] ι × κ →₀ M ⊗[R] N :=
+noncomputable def finsuppTensorFinsupp (R M N ι κ : Sort _) [CommSemiring R] [AddCommMonoid M]
+    [Module R M] [AddCommMonoid N] [Module R N] : (ι →₀ M) ⊗[R] (κ →₀ N) ≃ₗ[R] ι × κ →₀ M ⊗[R] N :=
```

Comment:
Can you make these generalizations in a separate PR, so that we can measure their performance impact? I think it was written this way in mathlib3 because it timed out otherwise.

Reply:
I can, but the rest of mathlib compiled swiftly here. (These are almost not used, I believe).

Reply:
done…